WIP: lmdb support

uhliarik opened the pull request:
There is still WIP.

uhliarik committed:
Initial commit for lmdb support

minfrin commented:
Not had a chance to review, but definite +1 in principle.

uhliarik commented:
Since I had to rebase this code to the trunk, I'm no longer able to run tests. If anyone can see where I'm missing linking the library, feel free to review the commit. 

```
$ make test
cd test && make all check
make[1]: Entering directory '/home/luhliari/playground/UPSTREAM/APR/apr_lmdb/test'
make[2]: Entering directory '/home/luhliari/playground/UPSTREAM/APR/apr_lmdb/test'
/bin/sh /home/luhliari/playground/UPSTREAM/APR/apr_lmdb/libtool --silent --mode=compile --tag=CC gcc -g -O2 -pthread   -DHAVE_CONFIG_H  -DLINUX -D_REENTRANT -D_GNU_SOURCE   -I../include -I./../include  -o globalmutexchild.lo -c globalmutexchild.c && touch globalmutexchild.lo
/bin/sh /home/luhliari/playground/UPSTREAM/APR/apr_lmdb/libtool --silent --mode=link --tag=CC gcc -g -O2 -pthread   -DHAVE_CONFIG_H  -DLINUX -D_REENTRANT -D_GNU_SOURCE   -I../include -I./../include   -no-install    -o globalmutexchild globalmutexchild.lo ../libapr-2.la   -lcrypt -lrt -lcrypt  -lpthread -ldl -lexpat
/usr/bin/ld: ../.libs/libapr-2.so: undefined reference to `mdb_drop'
/usr/bin/ld: ../.libs/libapr-2.so: undefined reference to `mdb_env_close'
/usr/bin/ld: ../.libs/libapr-2.so: undefined reference to `mdb_cursor_get'
/usr/bin/ld: ../.libs/libapr-2.so: undefined reference to `mdb_del'
/usr/bin/ld: ../.libs/libapr-2.so: undefined reference to `mdb_env_open'
/usr/bin/ld: ../.libs/libapr-2.so: undefined reference to `mdb_strerror'
/usr/bin/ld: ../.libs/libapr-2.so: undefined reference to `mdb_cursor_open'
/usr/bin/ld: ../.libs/libapr-2.so: undefined reference to `mdb_txn_commit'
/usr/bin/ld: ../.libs/libapr-2.so: undefined reference to `mdb_dbi_close'
/usr/bin/ld: ../.libs/libapr-2.so: undefined reference to `mdb_dbi_open'
/usr/bin/ld: ../.libs/libapr-2.so: undefined reference to `mdb_cursor_close'
/usr/bin/ld: ../.libs/libapr-2.so: undefined reference to `mdb_env_create'
/usr/bin/ld: ../.libs/libapr-2.so: undefined reference to `mdb_txn_begin'
/usr/bin/ld: ../.libs/libapr-2.so: undefined reference to `mdb_env_set_mapsize'
/usr/bin/ld: ../.libs/libapr-2.so: undefined reference to `mdb_put'
/usr/bin/ld: ../.libs/libapr-2.so: undefined reference to `mdb_get'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:125: globalmutexchild] Error 1
make[2]: Leaving directory '/home/luhliari/playground/UPSTREAM/APR/apr_lmdb/test'
make[1]: *** [/home/luhliari/playground/UPSTREAM/APR/apr_lmdb/build/apr_rules.mk:118: all-recursive] Error 1
make[1]: Leaving directory '/home/luhliari/playground/UPSTREAM/APR/apr_lmdb/test'
```

uhliarik committed:
Add -llmdb

Add -llmdb to test/Makefile.in so test can be compiled without
any error.

notroj commented on build.conf:
```diff
@@ -22,6 +22,7 @@ paths =
   crypto/crypt_blowfish.c
   dbm/apr_dbm_sdbm.c
   dbm/apr_dbm.c
+  dbm/apr_dbm_lmdb.c
```
This should be removed, the DBM drivers which can be built as DSOs should not be listed here otherwise it will get linked twice.

notroj commented:
With the line marked in my review above removed, and e147886e92b2858895fe9fff30fccf9dba678cad reverted, I can build with:

```sh
"./configure" \
"--with-lmdb" \
"--with-berkeley-db" \
"--with-dbm=db5" \
"--disable-modular-dso"
```

and this passes the tests. Thanks @uhliarik!

notroj committed:
Pass the -lock file back via _usednames. This seems to be reliable
if using MDB_NOSUBDIR per http://www.lmdb.tech/doc/group__mdb.html#ga32a193c6bf4d7d5c5d579e71f22e9340

notroj commented:
I've tested this with mod_dav and it seems to work well. As well as the changes described above I changed the `_usednames` implementation to pass back the "-lock" file as well:

https://github.com/notroj/apr/commit/26afe6b5b134ef3652bb629e6c68af38317f7f62

uhliarik committed:
Remove lmdb driver from build.conf and revert e147886

- Remove lmdb driver from build.conf so it won't get linked twice!
- Revert e147886e

uhliarik commented:
Thanks Joe for reviewing my changes. Previously, I was configurating the lmdb driver with:

```
./configure --with-lmdb --with-dbm=lmdb
```

And I was having troubles with loading lmdb driver as DSO. It was failing during DSO loading phase. I hope there is no issue with this and it is just some configuration thing.

uhliarik committed:
Removed wrong lines from build.conf in ad34f189

- Removed wrong lines from build.conf in ad34f189. Revering and
  removing correct line

uhliarik committed:
Pass the -lock file back via _usednames. This seems to be reliable

if using MDB_NOSUBDIR per http://www.lmdb.tech/doc/group__mdb.html#ga32a193c6bf4d7d5c5d579e71f22e9340

Merging the change suggested by @notroj
https://github.com/apache/apr/commit/26afe6b5b134ef3652bb629e6c68af38317f7f62

notroj committed:
Move #include <lmdb.h> inside conditional.

notroj committed:
Add LMDB test build.

notroj committed:
Use --with-lmdb for LMDB test build (duh).

notroj committed:
Fix Debian package name.

notroj committed:
Fix GCC warnings.

notroj committed:
Add BDB test.

notroj committed:
Remove unnecessary apr_dbm_t struct and add test for opening
a non-existent database.

notroj committed:
Simplify error handling - since mdb_* can return errno as positive
(just like apr_status_t) map those values directly; map all
other errors as APR_EGENERAL.

APR-internal APIs should not use error codes within the
APR_OS_START_USEERR number region (mistake common to dbm/*.c.)

notroj committed:
Default to using MDB_NOSYNC when opening the database to avoid
overhead of per-transaction fdatasync() calls, bringing comparable
behaviour/performance to BDB etc.

Arguably this behaviour could be controlled by a new flag passed
in the mode parameter of apr_dbm_open*.

uhliarik committed:
Merge pull request #1 from notroj/lmdb_support

Lmdb support

uhliarik committed:
Remove lmdb_retry function

which was originally used for dynamic database resizing in case the
database was full.

This function is no longer used and the size of the database is set
to UINT32_MAX in vt_lmdb_open() function.

notroj commented:
Thanks a lot for the contribution, @uhliarik
